src/tree.rs

```diff
@@ -135,3 +149,125 @@ pub fn minimum_spanning_tree(
 
     Ok(spanning_tree)
 }
+
+/// Bipartition tree by finding balanced cut edges of a spanning tree using
+/// node contraction. Assumes that the tree is connected and is a spanning tree.
+/// A balanced edge is defined as an edge that, when cut, will split the
+/// population of the tree into two connected subtrees that have population near
+/// the population target within some epsilon. The function returns a list of
+/// all such possible cuts, represented as the set of nodes in one
+/// partition/subtree.
+///
+/// :param PyGraph graph: Spanning tree. Must be fully connected
+/// :param pops: The populations assigned to each node in the graph.
+/// :param float pop_target: The population target to reach when partitioning the
+///     graph.
+/// :param float epsilon: The maximum percent deviation from the pop_target
+///     allowed while still being a valid balanced cut edge.
+///
+/// :returns: A list of tuples, with each tuple representing a distinct
+///     balanced edge that can be cut. The tuple contains the root of one of the
+///     two partitioned subtrees and the set of nodes making up that subtree.
+#[pyfunction]
+#[pyo3(text_signature = "(spanning_tree, pop, target_pop, epsilon)")]
+pub fn bipartition_tree(
+    _py: Python,
+    spanning_tree: &graph::PyGraph,
+    pops: Vec<f64>,
+    pop_target: f64,
+    epsilon: f64,
+) -> PyResult<Vec<(usize, Vec<usize>)>> {
```

Comment:
Hmm, what would a custom return type look like? Is this implemented in `retworkx` for other funcs?

Reply:
It's pretty easy to do you just add a macro call to `iterators.rs` to define the custom return type which just outlines the name, fields, types, docs, etc for each sequence type built from a vec. For example: https://github.com/Qiskit/retworkx/blob/main/src/iterators.rs#L619-L654 builds the custom `EdgeList` type from a `Vec<(usize, usize)>`. The only complication here is with the `(usize, Vec<usize>)` for each element you'd probably want to nest things so we don't do a copy on the inner vec too and instead return a tuple of `(usize, NodeIndices)` for each `__getitem__` call.

Reply:
Ok, if I were to do this:
```python
customvec[100]
```
would the `custom_vec_iter_impl` macro build up/calculate `customvec[0]`, `customvec[1]`, ...,`customvec[100]`? If so, I'm not sure how much performance improvement this will yield.

Reply:
It would only return the element at index 100. Basically the way it works is that it builds a `pyclass` struct that has a single `Vec` field. So if you did something like:
```rust
custom_vec_iter_impl!(
    customvec,
    inner_vec,
    (usize, Vec<usize>),
    "My Docstring"
);
```
that expands to something like:

```rust
#[doc = $doc]
#[pyclass(module = "retworkx")]
#[derive(Clone)]
pub struct customvec {
        pub inner_vec: Vec<(usize, Vec<usize>)>,
}
```

With the python sequence protocol implemented so `fn __getitem__(self, idx) -> (usize, Vec<usize>)` just is basically returning `self.inner_vec[idx]` and only does the conversion for that single element.

Reply:
As for the complexity I was briefly talking about before is that using `(usize, Vec<usize>)` as the data type will still do the iteration and conversion on the whole inner `Vec<usize>` inside the tuple. You can avoid that though by nesting custom returns so that you make the data type `(usize, NodeIndices)` instead of `(usize, Vec<usize>)`.

Reply:
Ah, I see. I'm implementing it now, then -- sounds like I should see a speedup with this method.

Reply:
Huh, I don't see a speedup (in fact, it gets slower). Strange.

Reply:
Hmm, interesting. That probably means that on the python side you're iterating over everything >=1 times. Doing custom return types like this is a tradeoff between returning fast and adding overhead on per element access.

Reply:
Yeah, but all I do is randomly pick (uniformly) one of the balanced edges. Maybe `random.choice` relies on `__len__` which forces it to be fully enumerated.

Reply:
Ah, it does use `__len__`: https://github.com/python/cpython/blob/75eee1d57eb28283a8682a660d9949afc89fd010/Lib/random.py#L371

It looks like `__len__` is automatically/efficiently implemented by the `custom_vec_iter_impl!` macro. This is even stranger.

Reply:
For context, this is how it's being used:
```python
    balanced_node_choices = retworkx.bipartition_graph_mst(
        graph.pygraph,
        lambda x: random.random(),
        pops,
        float(pop_target),
        float(epsilon)
    )
    balanced_nodes = {graph.retworkx_networkx_mapping[x] for x in choice(balanced_node_choices)[1]}
```

(https://github.com/mggg/GerryChain/pull/381/files?diff=unified&w=0#diff-02d17350b8ad584edb0ba9a458e607c2c9241895e2227bbbaf5d2580d83c52e1R193)

Reply:
There are some cases where there is only one (or zero) balanced edge, so this may be the explanation for the performance hit.